terminal: Implement OSC 7 (CWD, current directory)

glacambre opened the pull request:
This commit introduces a new 'autoshelldir' setting controlling whether
Neovim should aknowledge OSC7 sequences emitted by the programs running
in its terminal buffer.

The OSC7 sequence is sent by the child process to indicate that its
current working directory changed.

When Neovim receives OSC7, it stores the received path in a buffer-local
variable named "terminal_job_cwd", which is available to the user. It
also stores the path as the terminal buffer's name. This enables
'autochdir' to work as expected with buffer terminals.

Note that Vim also has a setting named 'autoshelldir' that does
something similar. This setting was implemented in patch 8.2.2675
(Vim commit 8b9abfd86c736ed3f298dfa8b50a962c118b3983). However, Vim's
approach has several drawbacks causing possible desynchronization
between the child process' cwd and the parent window. This is why this
patch isn't a port of Vim's feature, but a new implementation. However,
in the hope of sharing code with Vim in the future, an attempt was made
to keep function names and interfaces similar to Vim's. Moreover, the
url_decode function was imported as-is.

This closes #2252.

zeertzjq commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -298,6 +298,18 @@ describe(':terminal buffer', function()
+        parent = cwd:match("^(.+/)")
+    end
+    command('terminal')
+    feed('printf "\\033]7;file://host/' .. parent .. '..\\033\\\\"<CR>')
```
1. `printf` is shell-dependent
2. You feeded the keys in Normal mode

zeertzjq replied:
Looking at `test/functional/terminal/helpers.lua` it seems `chansend()` (a.k.a. `jobsend()`) should be used.

glacambre replied:
Thank you so much for your advice, it took me a while but thanks to it I managed to find the right sequence of bytes to make the test pass!

zeertzjq replied:
I doubt whether OSC 7 makes sense on Windows. I hardly know anything about terminal on Windows though.

glacambre replied:
I managed to fix the test on windows. Thanks again for the advice!

glacambre commented:
@zeertzjq is there anything else you think I should change about this PR? :)

zeertzjq reviewed (commented):
I'm not sure whether ignoring the hostname is the best way to deal with this, but I don't think there is an easy alternative either.

Changing `curbuf->b_ffname` may have some side-effects, like affecting the result of `getbufinfo()`, so I'm not sure about this either.

zeertzjq commented:
Another way to do this may be to replace the `{cwd}` part in `term://{cwd}//{pid}:{cmd}`, and extract the `{cwd}` part in `vim_chdirfile()`. This will not work for `nvim_open_term()` though.

zeertzjq commented:
I opened a PR to make `autochdir` extract CWD from `term://` URI: #16989

glacambre commented:
Thank you for this second round of review :).

>Changing curbuf->b_ffname may have some side-effects, like affecting the result of getbufinfo(), so I'm not sure about this either.

Right, it changes the "name" attribute returned by getbufinfo(). That seems like the expected behavior to me, could you describe the reason why you think this might be bad?

>Another way to do this may be to replace the {cwd} part in term://{cwd}//{pid}:{cmd}, and extract the {cwd} part in vim_chdirfile().

Great idea! I'll adapt my PR to produce `term://` url and see if your PR and mine interact properly.

zeertzjq committed:
feat(autochdir): extract CWD from term:// URI

glacambre committed:
feat(terminal): support OSC7 ('autoshelldir')

This commit introduces a new 'autoshelldir' setting controlling whether
Neovim should aknowledge OSC7 sequences emitted by the programs running
in its terminal buffer.

The OSC7 sequence is sent by the child process to indicate that its
current working directory changed.

When Neovim receives OSC7, it stores the received path in a buffer-local
variable named "terminal_job_cwd", which is available to the user. It
also stores the path as the terminal buffer's name. This enables
'autochdir' to work as expected with buffer terminals.

Note that Vim also has a setting named 'autoshelldir' that does
something similar. This setting was implemented in patch 8.2.2675
(Vim commit 8b9abfd86c736ed3f298dfa8b50a962c118b3983). However, Vim's
approach has several drawbacks causing possible desynchronization
between the child process' cwd and the parent window. This is why this
patch isn't a port of Vim's feature, but a new implementation. However,
in the hope of sharing code with Vim in the future, an attempt was made
to keep function names and interfaces similar to Vim's. Moreover, the
url_decode function was imported as-is.

glacambre commented:
I rebased this PR, along with @zeertzjq's commit, which is included in this branch. I also tried to make it match Vim's version of OSC7 where it makes sense. I'm happy with how it works now, but not 100% certain the code is optimal, I'll be glad to hear your reviews/comments :)

famiu commented on src/nvim/buffer.c:
```diff
@@ -2844,6 +2844,13 @@ void buf_name_changed(buf_T *buf)
   ml_timestamp(buf);            // reset timestamp
 }
 
+/// Generate a terminal buffer name of the form "term://$CWD//$PID:$CMD" in
+/// buf.
```
put the comment in a single line

glacambre replied:
Done, thanks!

famiu commented on src/nvim/terminal.c:
```diff
@@ -152,6 +154,111 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
+
+  String dir = cstr_to_string((char *)new_dir);
+  Error err = ERROR_INIT;
+  dict_set_var(curbuf->b_vars, cstr_as_string("terminal_job_cwd"),
```
```suggestion
  dict_set_var(curbuf->b_vars, STATIC_CSTR_AS_STRING("terminal_job_cwd"),
```

glacambre replied:
Done, thank you :)

justinmk commented on src/nvim/eval/funcs.c:
```diff
@@ -10785,21 +10785,30 @@ static void f_termopen(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   Error err = ERROR_INIT;
   // deprecated: use 'channel' buffer option
-  dict_set_var(curbuf->b_vars, cstr_as_string("terminal_job_id"),
+  dict_set_var(curbuf->b_vars, STATIC_CSTR_AS_STRING("terminal_job_id"),
```
these are deprecated (see above comment). instead the 'channel' buffer-local option holds the channel-id (job-id).

justinmk commented on src/nvim/eval/funcs.c:
```diff
@@ -10785,21 +10785,30 @@ static void f_termopen(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   api_clear_error(&err);
+  String command = cstr_to_string(cmd);
+  dict_set_var(curbuf->b_vars, STATIC_CSTR_AS_STRING("terminal_job_cmd"),
+               STRING_OBJ(command), false, false, &err);
```
we need a `jobget({id})` function, rather than storing this info in ad-hoc buffer vars.

justinmk commented on src/nvim/terminal.c:
```diff
@@ -152,6 +154,111 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
+///
+/// The OSC 7 sequence has the format of
+/// "\033]7;file://HOSTNAME/CURRENT/DIR\033\\"
+static void sync_shell_dir(const char *command, size_t cmdlen, void *user)
```
this is a lot of new C code. could be avoided if we had a generic OSC event or Lua callback: 

- https://github.com/neovim/neovim/issues/4413

glacambre replied:
Thanks for taking the time to review this PR!

>this is a lot of new C code. could be avoided (and more extensible) if we had a generic OSC event or Lua callback:

Yes, I kept thinking that making it a lua feature might be a better solution. I'll try to whip up a PR to help decide what the best approach is.

glacambre commented:
Closing this as there are lots of conflicts and https://github.com/neovim/neovim/pull/22159 seems like a better approach.
